Source/Diagnostics/ReducedDiags/RhoMaximum.cpp

```diff
@@ -0,0 +1,169 @@
+/* Copyright 2020 Neil Zaim
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "RhoMaximum.H"
+#include "Diagnostics/ComputeDiagFunctors/RhoFunctor.H"
+#include "WarpX.H"
+
+// constructor
+RhoMaximum::RhoMaximum (std::string rd_name)
+: ReducedDiags{rd_name}
+{
+
+    // RZ coordinate is not working
+#if (defined WARPX_DIM_RZ)
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(false,
+        "RhoMaximum reduced diagnostics does not work for RZ coordinate.");
+#endif
+
+    // read number of levels
+    int nLevel = 0;
+    amrex::ParmParse pp("amr");
+    pp.query("max_level", nLevel);
+    nLevel += 1;
+    m_rho_functors.resize(nLevel);
+
+    // We do not use coarsening in the functors for this diag.
+    const amrex::IntVect crse_ratio = amrex::IntVect(1);
+
+    // Initialize functors for the total charge density
+    for (int lev = 0; lev < nLevel; ++lev)
+    {
+        m_rho_functors[lev].push_back(std::make_unique<RhoFunctor>(lev, crse_ratio));
+    }
+
+    // get MultiParticleContainer class object
+    const auto & mypc = WarpX::GetInstance().GetPartContainer();
+
+    // get number of species (int)
+    const auto nSpecies = mypc.nSpecies();
+
+    // A vector to store the indices of charges species in mypc
+    amrex::Vector<int> indices_charged_species;
+
+    // number of charged species
+    int n_charged_species = 0;
+
+    for (int i = 0; i < nSpecies; ++i)
+    {
+        // Only charged species are relevant for this diag
+        if (mypc.GetParticleContainer(i).getCharge() != amrex::Real(0.))
+        {
+            indices_charged_species.push_back(i);
+            n_charged_species += 1;
+            for (int lev = 0; lev < nLevel; ++lev)
+            {
+                // Initialize functors for the charge density of each charged species
+                m_rho_functors[lev].push_back(std::make_unique<RhoFunctor>(lev, crse_ratio, i));
+            }
+        }
+    }
+
+    // get species names (std::vector<std::string>)
+    const auto species_names = mypc.GetSpeciesNames();
+
+    // Min and max of total rho + max of |rho| for each species
+    const int noutputs_per_level = 2+n_charged_species;
+    m_data.resize(nLevel*noutputs_per_level, amrex::Real(0.0));
+
+    if (amrex::ParallelDescriptor::IOProcessor())
+    {
+        if ( m_IsNotRestart )
+        {
+            // open file
+            std::ofstream ofs{m_path + m_rd_name + "." + m_extension,
+                std::ofstream::out | std::ofstream::app};
```

Comment:
Is this `std::ofstream::app` here for a reason? It seems to me that this function should write the first line of the output file...

Reply:
Yes I guess the append is superfluous, I'll remove it in all other reduced diags for consistency.